File: gateway/src/mcp/i18n.ts
import { FluentBundle, FluentResource } from "@fluent/bundle";

// FTL files are copied as .txt during prebuild so esbuild's built-in text loader handles them.
import enMcp from "../../locales/en/mcp.txt";
import plMcp from "../../locales/pl/mcp.txt";

function createBundle(locale: string, ftlText: string): FluentBundle {
  const bundle = new FluentBundle(locale);
  const resource = new FluentResource(ftlText);
  bundle.addResource(resource);
  return bundle;
}

const bundles: Record<string, FluentBundle> = {
  en: createBundle("en", enMcp),
  pl: createBundle("pl", plMcp),
};

/**
 * Translate a key to the given locale.
 * Falls back to "en" if the locale bundle doesn't have the key.
 * Falls back to the key itself if not found in any bundle.
 */
export function tr(key: string, locale: string): string {
  const bundle = bundles[locale] ?? bundles["en"];
  const msg = bundle.getMessage(key);
  if (msg?.value) {
    return bundle.formatPattern(msg.value);
  }
  // Fallback to English
  const enMsg = bundles["en"].getMessage(key);
  if (enMsg?.value) {
    return bundles["en"].formatPattern(enMsg.value);
  }
  return key;
}

File: crates/mcp/src/i18n.rs
use fluent_bundle::concurrent::FluentBundle;
use fluent_bundle::{FluentArgs, FluentResource, FluentValue};
use std::collections::HashMap;
use std::path::Path;
use unic_langid::LanguageIdentifier;

pub struct McpI18n {
    bundles: HashMap<String, FluentBundle<FluentResource>>,
}

impl McpI18n {
    /// Load FTL files from `<locales_dir>/<lang>/mcp.ftl`.
    pub fn load_from(locales_dir: &Path) -> Self {
        let mut bundles = HashMap::new();
        for lang in ["en", "pl"] {
            let path = locales_dir.join(lang).join("mcp.ftl");
            let src = std::fs::read_to_string(&path)
                .unwrap_or_else(|e| panic!("read {}: {e}", path.display()));
            let resource = FluentResource::try_new(src)
                .unwrap_or_else(|e| panic!("parse {}: {e:?}", path.display()));
            let id: LanguageIdentifier = lang.parse().expect("parse langid");
            let mut bundle = FluentBundle::new_concurrent(vec![id]);
            bundle.add_resource(resource).expect("add resource");
            bundle.set_use_isolating(false);
            bundles.insert(lang.to_string(), bundle);
        }
        Self { bundles }
    }

    /// Load FTL files from `locales/` relative to the current working directory.
    pub fn load() -> Self {
        Self::load_from(Path::new("locales"))
    }

    pub fn translate(&self, locale: &str, key: &str) -> String {
        self.translate_args(locale, key, &[])
    }

    pub fn translate_args(&self, locale: &str, key: &str, args: &[(&str, &str)]) -> String {
        let bundle = self
            .bundles
            .get(locale)
            .or_else(|| self.bundles.get("en"))
            .expect("en fallback bundle missing");
        let Some(msg) = bundle.get_message(key) else {
            return key.to_string();
        };
        let Some(pattern) = msg.value() else {
            return key.to_string();
        };
        let mut fluent_args = FluentArgs::new();
        for (k, v) in args {
            fluent_args.set(*k, FluentValue::from(v.to_string()));
        }
        let mut errors = vec![];
        bundle
            .format_pattern(pattern, Some(&fluent_args), &mut errors)
            .to_string()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn test_i18n() -> McpI18n {
        // CARGO_MANIFEST_DIR is crates/mcp; locales are two levels up
        let manifest = std::path::PathBuf::from(env!("CARGO_MANIFEST_DIR"));
        let locales = manifest.parent().unwrap().parent().unwrap().join("locales");
        McpI18n::load_from(&locales)
    }

    #[test]
    fn loads_en_and_pl() {
        let i = test_i18n();
        assert!(
            i.translate("en", "mcp-tool-create_item-desc")
                .contains("Create")
        );
        assert!(
            i.translate("pl", "mcp-tool-create_item-desc")
                .contains("Utwórz")
        );
    }

    #[test]
    fn formats_args() {
        let i = test_i18n();
        let s = i.translate_args("en", "mcp-err-not-found", &[("entity", "item")]);
        assert!(s.contains("item"), "got: {s}");
    }

    #[test]
    fn unknown_key_returns_key() {
        let i = test_i18n();
        assert_eq!(i.translate("en", "nonexistent-key"), "nonexistent-key");
    }
}
